File: tests/kernel.go
package main
import ("os"; "fmt")
func rot(n, x, y, rx, ry int) (int, int) {
    if ry == 0 { if rx == 1 { x, y = n-1-x, n-1-y }; return y, x }; return x, y
}
func d2xy(n, d int) (int, int) {
    x, y, t := 0, 0, d
    for s := 1; s < n; s *= 2 { rx := 1 & (t / 2); ry := 1 & (t ^ rx); x, y = rot(s, x, y, rx, ry); x += s * rx; y += s * ry; t /= 4 }
    return x, y
}
func main() {
    p := "vines_images/go_vine.ppm"; f, _ := os.Create(p); f.Write([]byte("P6\n4000 4000\n255\n")); f.Truncate(48000017); f.Close()
    f, _ = os.OpenFile(p, os.O_RDWR, 0644)
    for i := 121; i <= 160; i++ {
        sql := fmt.Sprintf("SQL_ID_%d_GO_VINE\x00", i); curD := 543210 + (i * 1000)
        for j := 0; j < len(sql); j += 3 {
            x, y := d2xy(4096, curD); end := j+3; if end > len(sql) { end = len(sql) }
            f.WriteAt([]byte(sql[j:end]), int64(17+(y*4000+x)*3)); curD++
        }
    }
}

File: tests/kernel.cpp
#include <iostream>
#include <fstream>
#include <string>
void rot(int n, int *x, int *y, int rx, int ry) {
    if (ry == 0) { if (rx == 1) { *x = n-1-*x; *y = n-1-*y; } int t = *x; *x = *y; *y = t; }
}
void d2xy(int n, int d, int &x, int &y) {
    int rx, ry, s, t=d; x = y = 0;
    for (s=1; s<n; s*=2) { rx = 1 & (t/2); ry = 1 & (t ^ rx); rot(s, &x, &y, rx, ry); x += s*rx; y += s*ry; t /= 4; }
}
int main() {
    std::string path = "vines_images/cpp_vine.ppm";
    std::ofstream out(path, std::ios::binary); out << "P6\n4000 4000\n255\n"; 
    out.seekp(48000016); out.put(0); out.close();
    std::fstream f(path, std::ios::in | std::ios::out | std::ios::binary);
    int h = 543210; // Shared Seed Logic
    for(int i=81; i<=120; i++){
        std::string sql = "SQL_ID_" + std::to_string(i) + "_CPP_VINE"; sql += '\0';
        int curD = h + (i * 1000);
        for(int j=0; j<sql.length(); j+=3){
            int x, y; d2xy(4096, curD, x, y);
            f.seekg(17 + (y*4000+x)*3); f.write(sql.substr(j,3).c_str(), 3);
            curD++;
        }
    }
    return 0;
}
